app/models/user_access.rb

```diff
@@ -79,4 +145,29 @@ def resources_for(resource_model, action)
 
     resource_model.where(id: resource_ids)
   end
+
+  def prepare_access_resources(selected_facility_ids)
+    selected_facilities = Facility.where(id: selected_facility_ids)
+    resources = []
+
+    selected_facilities.group_by(&:organization).each do |org, selected_facilities_in_org|
+      if can?(:manage, :organization, org) && org.facilities == selected_facilities_in_org
+        resources << {resource: org}
+        selected_facilities -= selected_facilities_in_org
+      end
+    end
+
+    selected_facilities.group_by(&:facility_group).each do |fg, selected_facilities_in_fg|
+      if can?(:manage, :facility_group, fg) && fg.facilities == selected_facilities_in_fg
+        resources << {resource: fg}
+        selected_facilities -= selected_facilities_in_fg
+      end
+    end
+
+    selected_facilities.each do |f|
+      resources << {resource: f} if can?(:manage, :facility, f)
+    end
+
+    resources.flatten
+  end
```

Comment:
This is where the magic happens. Love it.

Reply:
I would like to re-write this a little more _functionally_, but it works pretty well for now, so I've kept it around.